### PR TITLE
Add evaluation of dataset expressions

### PR DESCRIPTION
**Pull request by daniel-thom**
## Description

Abstracts SQL operations to combine datasets with a mathematical expression syntax. Clones the repository https://github.com/axiacore/py-expression-eval and modifies it to support a `union` operation with `|`.

### CODE REVIEW

**Review by elainethale — COMMENTED** [review]
I'm good with the syntax

File: tests/test_dataset_expression_handler.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+def test_dataset_expression_union(datasets):
+    df = evaluate_expression("dataset1 | dataset2", datasets).df.cache()
+    assert df.count() == 6
+    assert df.filter("county == 'Jefferson'").count() == 2
+    assert df.filter("county == 'Boulder'").count() == 2
+    assert df.filter("county == 'Denver'").count() == 2
+
+
+def test_dataset_expression_combo(datasets):
```
Comment (lixiliu) [inline tests/test_dataset_expression_handler.py:64]: should we also test for invalid cases too, i.e., where columns or size do not match, duplicated stacked dimensions? I know the latter is more related to dsgrid and not this added feature, so maybe punt the latter to relevant PR?

**Review by lixiliu — APPROVED** [review]
Minor fixes and suggestions

File: dsgrid/dataset/dataset_expression_handler.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        self.df = df
+        self.dimension_columns = dimension_columns
+        self.pivoted_columns = pivoted_columns
+
+    def _op(self, other, op):
+        orig_self_count = self.df.count()
+        orig_other_count = other.df.count()
+        if orig_self_count != orig_other_count:
+            raise DSGInvalidOperation(
+                "operation {op} requires that the datasets have the same length "
```
[the remark this answers is not in the document]
Comment (daniel-thom) [reply dsgrid/dataset/dataset_expression_handler.py]: Thanks. All are fixed.

File: dsgrid/utils/py_expression_eval/tests.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        self.assertEqual(parser.parse("'x'=='x'").evaluate({}), True)
+        self.assertEqual(parser.parse("(a+b)==c").evaluate({'a': 1, 'b': 2, 'c': 3}), True)
+        self.assertEqual(parser.parse("(a+b)!=c").evaluate({'a': 1, 'b': 2, 'c': 3}), False)
+        self.assertEqual(parser.parse("(a^2-b^2)==((a+b)*(a-b))").evaluate({'a': 4859, 'b': 13150}), True)
+        self.assertEqual(parser.parse("(a^2-b^2+1)==((a+b)*(a-b))").evaluate({'a': 4859, 'b': 13150}), False)
+        self.assertEqual(parser.parse("(a**2-b**2)==((a+b)*(a-b))").evaluate({'a': 4859, 'b': 13150}), True)
+        self.assertEqual(parser.parse("(a**2-b**2+1)==((a+b)*(a-b))").evaluate({'a': 4859, 'b': 13150}), False)
+        self.assertExactEqual(parser.parse("x/((x+y))").simplify({}).evaluate({'x': 1, 'y': 1}), 0.5)
+        self.assertExactEqual(parser.parse('origin+2.0').evaluate({'origin': 1.0}), 3.0)
+        assert parser.parse('a | b').evaluate({'a': set([1, 2]), 'b': set([3, 4])}) == set([1, 2, 3, 4])
```
[the remark this answers is not in the document]
Comment (daniel-thom) [reply dsgrid/utils/py_expression_eval/tests.py:77]: I did add a test.

**Pull request merged**